content/blog/2020-04-preemption.md

```diff
@@ -0,0 +1,183 @@
+inject any sort of yield point.
+
+## Per-task operation budget
+
+Even though Tokio is not able to **preempt**, there is still an opportunity to
+nudge a task to yield back to the scheduler. As of [0.2.14], each Tokio task has
+an operation budget. This budget is reset when the scheduler switches to the
+task. Each Tokio resource (socket, timer, channel, ...) is now aware of this
+budget. As long as the task as budget remaining, the resource operates as it did
+previously. Each asynchronous operation (actions that users must `.await` on)
+decrement the task's budget. Once the task is out of budget, all resources will
+become "not ready" until the task yields back to the scheduler, at which point,
+the budget is reset.
+
+Going back to the echo server example from above. When the task is scheduled, it
+is assigned a budget of 128 operations. When `socket.read(..)` and
+`socket.write(..)` are called, the budget is decremented. If the budget is zero,
+the task yields back to the scheduler. If either `read` or `write` cannot
+proceed due to the underlying socket not being ready (no pending data or a full
+send buffer), then the task also yield back to the scheduler.
+
+The idea originated from a conversation I was having with [Ryan Dahl][ry]. He is
+using Tokio as the underlying runtime for [Deno][deno]. When doing some HTTP
+experimentation (a while back) with [Hyper], he was seeing some high tail
+latencies in some benchmarks. The problem was due to a loop not yielding back to
+the scheduler under load. Hyper ended up fixing the problem by hand in this one
+case, but Ryan mentioned that, when he worked on [node.js][node], they handled
+the problem by adding **per resource** limits. So, if a TCP socket was always
+ready, it would force a yield every so often. I mentioned this conversation to
+[Jon Gjenset][jonhoo], and he ended up coming up with the idea of placing the
+limit on the task itself instead of on each resource.
+
+The end result is that Tokio should be able to provide more consistent runtime
+behavior under load. While the exact heuristics will most likely be tweaked over
+time, initial measurements show that, in some cases, tail latencies are reduced
+almost 3x.
+
+[![benchmark](https://user-images.githubusercontent.com/176295/73222456-4a103300-4131-11ea-9131-4e437ecb9a04.png)](https://user-images.githubusercontent.com/176295/73222456-4a103300-4131-11ea-9131-4e437ecb9a04.png)
+
+"master" is before the automatic yielding and "preempt" is after. Click for a
+bigger version, also see the original [PR comment][pr] for more details.
+
+## A note on blocking
+
+While automatic cooperative task yielding improves many cases, it cannot preempt
+tasks. Users of Tokio must still take care to avoid CPU intensive work and using
+blocking APIs. The [`spawn_blocking`][spawn_blocking] function can be used to
+"asyncify" these sorts of tasks by running them on a thread pool.
```

Comment:
Do we also want to mention `block_in_place` here?

Reply:
I left it out mostly because I would rather users default to `spawn_blocking`. I can add it though.

Reply:
Either way is fine by me.